### ISSUE TITLE
aten.native_dropout: train=None is interpreted inconsistently across backends, AD modes, and compile/export paths

### ISSUE THREAD

**Issue by sha7doww** [pytorch/pytorch#197846]
### 🐛 Describe the bug

`native_dropout`'s schema is `native_dropout(Tensor input, float p, bool? train)` — `train` is optional with no
default, so an explicit `None` is accepted. I could not find documentation defining what `None` means. In
practice the stack does not agree: some consumers treat it as training, others as evaluation, and one throws.

**1. CPU and MPS eager kernels disagree**

```python
import torch

for p in (0.0, 0.5, 1.0):
    for dev in ("cpu", "mps"):
        y, m = torch.ops.aten.native_dropout.default(torch.ones(64, device=dev), p, None)
        print(f"p={p} {dev:4s} values={sorted(set(y.cpu().tolist()))} mask_all_true={bool(m.cpu().all())}")
```

```text
p=0.0 cpu  values=[1.0]      mask_all_true=True
p=0.0 mps  values=[1.0]      mask_all_true=True
p=0.5 cpu  values=[0.0, 2.0] mask_all_true=False
p=0.5 mps  values=[1.0]      mask_all_true=True
p=1.0 cpu  values=[0.0]      mask_all_true=False
p=1.0 mps  values=[1.0]      mask_all_true=True
```

At `p=1` CPU zeroes the tensor and MPS returns it unchanged. `native_dropout_mps` in
`aten/src/ATen/native/mps/operations/Dropout.mm` resolves the optional with `value_or(false)`, whereas
`native_dropout_cpu` (and the CUDA and XPU kernels) take the evaluation shortcut only for an *engaged* false.

**2. `torch.compile` disagrees with eager**

```python
import torch

def f(x):
    return torch.ops.aten.native_dropout.default(x, 0.5, None)[0]

x = torch.ones(256)
print("eager:   ", f(x).unique().tolist())
print("inductor:", torch.compile(f, backend="inductor", fullgraph=True)(x).unique().tolist())
```

```text
eager:    [0.0, 2.0]
inductor: [1.0]
```

`backend="eager"` and `"aot_eager"` keep dropout; `torch._inductor.config.fallback_random = True` also restores
it. The cause is `torch/_decomp/decompositions.py`'s `native_dropout`, whose guard `if train and p != 0` treats
`None` as falsy; inductor selects it through `extra_random_decomps` → `fast_random_decomps()` →
`select_decomp_table()` → `compile_fx`. Replacing only that entry with one that reads `None` as training makes
the compiled forward drop, and restoring it brings the identity back.

**3. Forward-mode and reverse-mode AD disagree, on one invocation**

```python
import torch
from torch.autograd.forward_ad import dual_level, make_dual, unpack_dual

x = torch.ones(256, dtype=torch.float64, requires_grad=True)
v = torch.arange(1, 257, dtype=x.dtype)

with dual_level():
    out, mask = torch.ops.aten.native_dropout.default(make_dual(x, v), 0.5, None)
    y, jvp = unpack_dual(out)

y.backward(v)
print("forward:", y[mask].unique().tolist())
print("JVP / v:", (jvp[mask] / v[mask]).unique().tolist())
print("VJP / v:", (x.grad[mask] / v[mask]).unique().tolist())
```

```text
forward: [2.0]
JVP / v: [2.0]
VJP / v: [1.0]
```

In `tools/autograd/derivatives.yaml`, the forward-mode rule applies the scale when
`!train.has_value() || train.value()`, while the reverse rule applies it only when
`train.has_value() && train.value()`. Because those rules are shared codegen but the kernels are not, a
*different* AD mode is wrong on each backend: on CPU the reverse rule disagrees with the forward, and on MPS
the forward-mode rule does.

**4. TorchScript raises on a constant `None`**

`Node::isNondeterministic()` converts the `train` argument to bool unconditionally, so a scripted
`aten::native_dropout(..., None)` node raises `expected bool`. This also affects `torch.jit.freeze`, and it is
what a legacy ONNX export hits first, during the CSE pass that calls `isNondeterministic()`.

### Proposed resolution

Standardise explicit `None` to **training**, matching CPU, CUDA and XPU, the forward-mode AD rule, and the
functorch batching rule, and correct the consumers that disagree. Alternatives are worse: giving `train` a
default only affects an *omitted* argument, not an explicit `None`; making `None` mean evaluation would change
working CPU/CUDA/XPU forwards and require changing forward AD and the batching rule too; rejecting `None` would
remove an input the schema admits.

Core fix — these three belong together, since the first two alone would regress MPS:

- `tools/autograd/derivatives.yaml`, reverse rule: change both copies of the scale condition from
  `!train.has_value() || !train.value()` to `train.has_value() && !train.value()`, keeping the `p == 1`
  handling. This is exactly the negation of the adjacent forward-mode predicate, so the two lines then encode
  one decision.
- `torch/_decomp/decompositions.py`, `native_dropout`: `if train and p != 0:` → `if (train is None or train) and p != 0:`
- `aten/src/ATen/native/mps/operations/Dropout.mm`, `native_dropout_mps`: `train.value_or(false)` →
  `train.value_or(true)`.

Companion fixes, each independent of the above:

- `Node::isNondeterministic()`: retrieve an optional IValue rather than forcing bool, preserving an explicit
  `None` and keeping the op conservatively nondeterministic.
- Legacy ONNX (`symbolic_opset12.py`): normalise `None` in the `native_dropout` symbolic before calling the
  shared helper, rather than changing the helper's non-optional contract.
- Modern ONNX: the ONNXScript translator takes `train: bool = True`, which an explicit `None` does not
  activate, and an unspecified ONNX `training_mode` means false. Since an op with an ONNX translation is
  excluded from the ATen decomposition table, fixing the decomposition does not cover this route.

### Tests

A dedicated eager-AD test over `train ∈ {None, False, True}` × `p ∈ {0, 0.2, 0.5, 1}`, asserting forward, JVP,
ordinary VJP, the differentiable-backward branch and a nonlinear-loss Hessian against the returned mask and the
canonical scale. `test_mps.py`'s `test_dropout` should gain `None` *and* a corrected oracle — it currently
branches on `if train`, which would encode the same wrong interpretation. Inductor and functionalized-RNG need
their own cases, asserting output values against the returned mask rather than comparing random draws.

Note two existing barriers: `test_decomp.py` both ignores `aten.native_dropout.default` and bypasses
`nondeterministic_seeded` ops, and `sample_inputs_dropout` uses `train=None` as a generator sentinel meaning
"emit both booleans", so it never produces an explicit `None` sample.

### Versions

```text
PyTorch version: 2.14.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A
OS: macOS 26.5 (arm64)
Clang version: 21.0.0 (clang-2100.1.1.101)
Python version: 3.12.13 (64-bit runtime)
Python platform: macOS-26.5-arm64-arm-64bit
Is CUDA available: False
CPU: Apple M4 Pro
```

CPU and MPS results above are measured on 2.14.0; the compile and AD behaviour also reproduces on 2.10.0+cpu
(`449b1768410104d3ed79d3bcfe4ba1d65c7f22c0`). CUDA and XPU claims are from source reading only — I have no such
device. The modern-ONNX claim is source-traced, not executed.

cc @justinchuby @titaiwangms @albanD @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20 @chauhang @penguinwu

**Comment by AnkitNakhawa** [comment]
Hi, I’ve submitted PR #197854 for the `native_dropout(train=None)` inconsistency in #197846. Could a maintainer assess whether the issue is actionable and approve the pending MPS/Inductor CI runs? The PR includes regression tests, but native validation is still pending. Thanks!
